Feature/add bot wallet balance alarm

krisoshea-eth opened the pull request:
feat: adds bot wallet balance alarm 

**Motivation**

This PR was created in order to implement a bot wallet balance alarm in the bot script. This alarm will notify the team if the bot wallet balance falls below the predefined healthy threshold which is calculated using the following formula:  Wallet balance = minSponsorTokens * Number of open positions.

**Summary**

In order to achieve this, changes were made to the bot's index.js file which is run during the bot start-up process. The changes made to the index.js file allow for the bot wallet balance alarm to be created when the bot script is run.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ X]  Ran end-to-end test, running the code as in production

krisoshea-eth committed:
Update index.js to include bot wallet balance check

This commit adds an additional check to the index.js script for the liquidator bot such that the script checks if the liquidator bot wallet balance is about the healthy balance threshold (minSponsorTokens * number of open positions)

krisoshea-eth committed:
Update Index.js

This commit adds crucial changes to the previous commit to ensure that the script outputs the correct values for the bot wallet balance

KohliSuraj commented on packages/liquidator/index.js:
```diff
@@ -155,6 +155,57 @@ async function run({
+        this.web3 = this.client.web3;
+        this.syntheticToken = syntheticToken;
+
+        await this.client.update();
```
we don't really need to update `financialContractClient` here again. Since, liquidator.update updates the financialContractClient too.

KohliSuraj commented on packages/liquidator/index.js:
```diff
@@ -155,6 +155,57 @@ async function run({
+
+        const numOfOpenPositions = this.client.getAllPositions().length;
+
+        console.log(`minSponsorTokens: ${minSponsorTokens}`);
```
I think we can use the variables already declared and fetched by this script previously, instead of fetching again?

krisoshea-eth committed:
Amend index.js based on feedback

This commit amends the original changes to the index.js file based on the feedback provided. Specifically, I removed the financialCliantContract update function from the calculateBotWalletBalance function and also removed the initialization of the minSponsorTokens variable as it is already initialised earlier in the script.

KohliSuraj committed:
refactoring wallet balance alarm

KohliSuraj committed:
minor change

KohliSuraj committed:
Merge branch 'feature/add-bot-wallet-balance-alarm' of github.com:SumeroApp/uma-protocol into feature/add-bot-wallet-balance-alarm

KohliSuraj committed:
remove extra comment

KohliSuraj committed:
fixing incorrect merge conflict change

KohliSuraj committed:
putting back the comment

krisoshea-eth committed:
Update Wallet Balance Alarm

This commit updates the bot wallet balance alarms to include collateral balance alarms for the disputer bots and also amends the original liquidator bot alarm to include the ooReward minimum amount required to be available in the liquidator bot wallet balance at all times

krisoshea-eth committed:
Update index.js

This commit updates the disputer index.js file to include the correct path for the walletBalanceAlarm.js file

KohliSuraj commented on packages/disputer/index.js:
```diff
@@ -115,10 +116,8 @@ async function run({
-      getVersionedAbi(disputerConfig.contractType),
-      financialContractAddress
-    );
+    const abi = getVersionedAbi(disputerConfig.contractType);
```
we can revert this change, seems unnecessary

KohliSuraj commented on packages/liquidator/src/walletBalanceAlarm.js:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  async calculateTargetLiquidatorSynthBalance() {
+    this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;
```
We should move `this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;` to the constructor, so that we can use this value throughout the class.

krisoshea-eth replied:
I don't think this is possible as we need to update the value of this.numOfOpenPositions regularly. If we move it to the constructor, it will always use the original value fetched in the constructor.

KohliSuraj replied:
@krisoshea-eth I agree with you 👍. Your approach is correct.

KohliSuraj commented on packages/liquidator/src/walletBalanceAlarm.js:
```diff
@@ -0,0 +1,157 @@
+
+  async calculateTargetLiquidatorCollateralBalance() {
+    const ooReward = await this.financialContract.methods.ooReward().call();
+    this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;
```
reuse `this.numOfOpenPositions`

KohliSuraj commented on packages/liquidator/src/walletBalanceAlarm.js:
```diff
@@ -0,0 +1,157 @@
+
+  async calculateUsdcTargetBalance() {
+    const allPositions = await this.financialContractClient.getAllPositions();
+    this.totalCollateralAmount = await allPositions.reduce((total, position) => {
```
I think we can directly use `totalPositionCollateral` from the contract, instead of recalculating here.

krisoshea-eth replied:
Good point! I didn't realise you could read that directly from the contract.

KohliSuraj commented on packages/liquidator/src/walletBalanceAlarm.js:
```diff
@@ -0,0 +1,157 @@
+    this.disputerBondPercentage = await this.financialContract.methods.disputeBondPercentage().call();
+    console.log(`Dispute Bond Percentage: ${this.disputerBondPercentage / 1e18}`);
+
+    const usdcTargetBalance =
```
This looks correct ✅

KohliSuraj replied:
I would just add extra brackets, to make sure correct order of calculation is being followed.

(this.totalCollateralAmount * (this.disputerBondPercentage / 1e18)) + ((this.numOfOpenPositions * this.bufferPercentage) * ooReward);

krisoshea-eth replied:
For some reason, everytime I push this change it reverts to the original version, even after I have disabled Prettier

KohliSuraj commented on packages/liquidator/src/walletBalanceAlarm.js:
@krisoshea-eth Can you move the disputer logic to a separate file under **disputer/src** similar to what we do for liquidator? It'll be much more readable and cleaner.

Probably, I didn't mention this explicitly before, but I wanted you to create a new class for disputer similar to walletBalanceAlarm.js in liquidator.

krisoshea-eth committed:
Update Bot Balance Alarms

This commit allows for the creation of a new class for both the disputer bot and liquidator bot when the index.js file is run. This commit also adds all of the recommended amendments provided by Suraj's feedback.

krisoshea-eth committed:
Update disputerBalanceAlarm.js

This commit adds a minor update to the calculateUsdcTargetBalance() function.

krisoshea-eth committed:
Update disputerBalanceAlarm.js

This commit adds a minor update to the calculateUsdcTargetBalance() function

KohliSuraj commented:
Closing in favour of #31
